src/node-debug.js

```diff
@@ -0,0 +1,154 @@
+var EventEmitter   = require('events');
+
+const delay = ms => new Promise(r => setTimeout(r, ms));
+
+const RETRY_DELAY     = 300;
+const MAX_RETRY_COUNT = 10;
+
+module.exports = class NodeDebug {
+    constructor (port = 5858, host = '127.0.0.1') {
+        this.currentPacketNumber = 1;
+        this.events              = new EventEmitter();
+        this.port                = port;
+        this.host                = host;
+        this.socket              = new Socket();
+        this.buffer              = Buffer.alloc(0);
+        this.getPacketPromise    = Promise.resolve();
+        this.sendPacketPromise   = Promise.resolve();
+
+        this.nodeInfo = {
+            v8Version:       '',
+            protocolVersion: '',
+            embeddingHost:   ''
+        };
+    }
+
+    async _attemptToConnect (port, host) {
+        this.socket.connect(port, host);
+
+        var connectionPromise = Promise.race([
+            promisifyEvent(this.socket, 'connect'),
+            promisifyEvent(this.socket, 'error')
+        ]);
+
+        return await connectionPromise
+            .then(() => true)
+            .catch(() => delay(RETRY_DELAY));
+    }
+
+    async _connectSocket (port, host) {
+        var connected = await this._attemptToConnect(port, host);
+
+        for (var i = 0; !connected && i < MAX_RETRY_COUNT; i++)
+            connected = await this._attemptToConnect(port, host);
+
+        if (!connected)
+            throw new Error('Unable to connect');
+
+        this.socket.on('data', data => this._handleNewData(data));
+    }
+
+    async _writeSocket (message) {
+        if (!this.socket.write(message))
+            await promisifyEvent(this.socket, 'drain');
+    }
+
+    _handleNewData (data) {
+        this.buffer = Buffer.concat([this.buffer, data]);
+
+        this.events.emit('new-data');
+    }
+
+    _getPacket () {
+        this.getPacketPromise = this.getPacketPromise.then(async () => {
+            var index = this.buffer.indexOf('\r\n\r\n');
+
+            while (index < 0) {
+                await promisifyEvent(this.events, 'new-data');
+
+                index = this.buffer.indexOf('\r\n\r\n');
+            }
+
+            var packet = {
+                headers: null,
+                body:    null
+            };
+
+            packet.headers = this.buffer
+                .toString('utf8', 0, index)
+                .split('\r\n')
+                .map(line => line.match(/^([^:]+):\s+(.*)$/))
+                .reduce((obj, match) => {
+                    obj[match[1].toLowerCase()] = match[2];
+
+                    return obj;
+                }, {});
+
+            var contentLength = packet.headers['content-length'] && parseInt(packet.headers['content-length'], 10);
+
+            if (!contentLength) {
+                this.buffer = this.buffer.slice(index + 4);
+                return packet;
+            }
+
+
+            while (this.buffer.length - index - 4 < contentLength)
```

Comment:
I would refactor all low level logic, because `index - 4`, `index + 4` doesn't have any sense  for me.